cpp/src/fil/common.cuh

```diff
@@ -108,17 +105,36 @@ struct sparse_storage {
 typedef sparse_storage<sparse_node16> sparse_storage16;
 typedef sparse_storage<sparse_node8> sparse_storage8;
 
+/// all model parameters mostly required to compute shared memory footprint,
+/// also the footprint itself
+struct shmem_size_params {
+  /// the most shared memory a kernel can request on the GPU in question
+  int max_shm;
+  /// for class probabilities, this is the number of classes considered;
+  /// num_classes is ignored otherwise
+  int num_classes;
+  /// how many columns an input row has
+  int num_cols;
+  /// are the input columns are prefetched into shared
+  /// memory before inferring the row in question
+  bool cols_in_shmem;
+  /// n_items is the most items per thread that fit into shared memory
+  int n_items;
+  /// shm_sz is the associated shared memory footprint
+  int shm_sz;
+};
+
+template <leaf_algo_t leaf_algo, int NITEMS>
+void try_nitems(int* num_items, size_t* shm_sz, shmem_size_params params);
+
 // predict_params are parameters for prediction
 struct predict_params {
```

Comment:
Consider using inheritance here: `struct predict_params : public shmem_size_params { ... };`. This will make the code cleaner, and will avoid a lot of unnecessary changes.